Hip-1193 -  block stream cutover (#1473)
hiero-ledger/hiero-improvement-proposals: HIP/hip-1193.md

Reviewer: ```suggestion
1. Verification of logic for delivering block streams.
2. Verification of block production and correctness.
3. Verification of ingestion and cloud storage.
4. Verification of block proof creation and verification before and after node upgrade.
5. Verification of Mirror Node transition detection and cutover capability.
```

Reviewer 2: Can we add a link to 1427?

Author: Yes, good call.

requires: 1056
replaces: 679
created: 2025-05-08
updated: 2026-05-06
---

## Abstract

This HIP defines the requirements and implementation details for transitioning the Hedera network
from record and event streams to block streams. This transition is a critical step toward enhancing
blockchain compatibility, improving data integrity, and enabling future network optimizations such
as state proofs.

The transition is structured as a **two-stage cutover**:

> - **Stage 1 — WRB Cutover**: Consensus nodes begin producing Wrapped Record Blocks (WRBs) and

Reviewer 2: Can we link to 1427 and discuss somewhere that the WRB format is discussed in that HIP.

Author: Yep, good call.

- **Network Scalability** - Block nodes assume responsibility for state-related services—including
  node reconnection (re-sync), state proof generation, and state backups—thereby relieving consensus
  nodes of these duties and allowing them to dedicate their full resources to transaction processing
  and consensus.
- **Reduced Transition Risk** — A two-stage cutover allows the block stream infrastructure to be
  validated under real network conditions before the irreversible cryptographic cutover to TSS
  occurs, significantly reducing the risk of data loss or service disruption.

Given the scale and interconnected nature of these changes, a structured migration plan is essential
to avoid disruption for mirror nodes and other downstream consumers that currently depend on legacy
record streams.

## Rationale

The transition from record streams to block streams requires careful planning to ensure network
continuity, data integrity, and backward compatibility. Block stream data will be available both
from block nodes and limited history cloud buckets. This HIP proposes a clean cutover approach with
the following key design decisions:

### Two-Stage Cutover for Risk Mitigation

> The cutover is divided into two discrete stages to minimize risk:

Reviewer: We lack a definition of terms used in the next paragraph, so perhaps include a bit more introduction here.

```suggestion
The cutover is divided into two discrete stages to minimize risk. The first stage relies on a modified form of the familiar record file data that we call a Wrapped Recordfile Block (WRB). The second stage completes the transition with cryptographic upgrades with the Threshold Signature Scheme (TSS).
```

**Stage 1 (WRB Cutover)** serves as a live integration test of the block stream infrastructure.
Consensus nodes produce WRBs in parallel with record streams. Record streams remain the
authoritative cryptographic source — if bugs or issues are discovered in WRB production or
streaming, consensus nodes continue producing record files unaffected. This stage allows block
nodes, mirror nodes, and downstream consumers to validate the new format and delivery pipeline
without any cryptographic or operational commitment.

**Stage 2 (TSS Cutover)** is the full, irreversible cryptographic cutover. This stage only
proceeds once Stage 1 has met defined readiness criteria, including successful WRB streaming
to block nodes, successful mirror node ingestion of WRBs, and verified block proof correctness.
At this point, record and event stream production ceases and TSS-signed block streams become

> the sole authoritative source of network data.

Reviewer: Should we include criteria regarding TSS quality and capability?

```suggestion
**Stage 2 (TSS Cutover)** is the full, irreversible cryptographic cutover. This stage only
proceeds once Stage 1 has met defined readiness criteria, including successful WRB streaming
to block nodes, successful mirror node ingestion of WRBs, and verified block proof correctness.
During this time we will also stress-test the TSS implementation and ensure that the new
cryptography and signature processes function correctly across all functionality and
performance ranges. At this point, record and event stream production ceases and TSS-signed
block streams become the sole authoritative source of network data.
```

This staged approach separates the infrastructure risk (can we produce and deliver block streams

> correctly?) from the cryptographic risk (can we permanently commit to TSS-signed blocks?), allowing

Reviewer 2: I would say this is less of a question around delivering block streams and more around block nodes since it's not the full BS format.

- Prevents data loss during the transition.
- Eases mirror node operations, enabling mirror nodes to update the mirror node version before the
  block stream cutover and have seamless service continuity.

During Stage 1, mirror nodes that are configured to receive WRBs will automatically fall back to
record streams if WRBs cannot be received or validated. This fallback is automatic and requires no
operator intervention.

### Dual Publishing and Cloud Storage

Block streams are accessed by mirror nodes from both block nodes and cloud storage buckets, to
provide redundancy and high availability. Block nodes are the primary delivery mechanism, with
cloud storage serving as a fallback if a block node connection is unavailable or fails. Mirror
nodes will still require a software upgrade to support the block stream format regardless of which
source they retrieve data from.

New cloud storage buckets will be created to hold this block stream data. These buckets are
limited-history and contain only the most recent data. Syncing data from genesis will require
connecting to block nodes or importing a state snapshot. New buckets are created rather than

> reusing existing ones to allow retention policies to efficiently prune historical data.

Reviewer: ```suggestion
New cloud storage buckets will be created to distribute this block stream data. These buckets are
limited-retention and contain only the most recent data. Syncing data from genesis will require
connecting to block nodes or importing a state snapshot. New buckets are created rather than
reusing existing ones to allow retention policies to efficiently prune historical data.
```

- Mirror nodes will receive this transaction and persist the ledger ID in their respective
  databases.
- Mirror node code will be updated to hardcode this ledger ID per network for any operators that
  start after this ledger ID publication.

### 2\. Cutover Point Definition

The transition from record streams to block streams occurs in two discrete stages.

#### 2.1 Stage 1 — WRB Cutover

The WRB Cutover is a live integration test of the block stream pipeline. The following conditions
define this stage:

- Consensus nodes begin producing WRBs alongside continuing record stream production.
- WRBs are streamed to a limited, designated set of block nodes for integration testing.
- **Record streams continue to be produced and remain the cryptographic authority of the network.**
  Consensus nodes will not halt record file production due to bugs or issues with WRB streaming.
- Mirror nodes may be optionally configured to receive WRBs during this period.
- Mirror nodes configured to receive WRBs will automatically revert to record streams if WRBs
  cannot be received or fail validation.
- This stage does not constitute a cryptographic cutover. No TSS signing of blocks occurs as the

>   authoritative proof during this stage.

Reviewer: ```suggestion
- This stage does not constitute a cryptographic cutover. No TSS signing of blocks occurs during this stage.
```

Stage 1 concludes and Stage 2 is authorized only when all of the following readiness criteria
are met:

- WRBs are successfully streamed to and stored by designated block nodes without errors.
- At least one mirror node has successfully ingested and validated WRBs end-to-end.
- Block proof correctness has been verified for a sustained period.
- No critical defects remain open against WRB production or delivery.

> 

Reviewer: Perhaps add a TSS readiness criteria here?

```suggestion
- TSS and WRAPS processing is fully tested and validated
- TSS and WRAPS meets all relevant performance and correctness criteria.

```

#### 2.2 Stage 2 — TSS Cutover

The TSS Cutover is the full cryptographic cutover from record streams to block streams. This stage
proceeds only after Stage 1 readiness criteria are satisfied, and occurs at a network update
boundary:

- Occurs after a Freeze Upgrade transaction during the mainnet update window before transition.
- The current round comes to consensus.
- The final record stream will be completed.

> - WRB generated from the final record in the record stream.

Reviewer: This might actually be produced _before_ the network upgrade.
I think the team is working to make that possible.

- First TSS-signed block of block stream produced.

### 3\. Consensus Node

#### 3.1 Wrapped Record Blocks for Historical Data

To allow migration to a single standardized data format while fully preserving data integrity and
cryptographic verifiability, historical data — including legacy record files, sidecars, and their
associated signature files — will be wrapped into Wrapped Record Blocks (WRBs) during the
transition period.

Each WRB is constructed by:

* Embedding the original record stream contents within a Block Item.

> * Appending the required block header (including consensus round metadata and timestamps) and
>   footer.

Reviewer: ```suggestion
* Prepending the required block header (including consensus round metadata and timestamps) and
  appending a block footer with hash values required to make verification self-contained.
```

> * Generating a cryptographic block proof derived from the individual node RSA signatures
>   originally produced for that record file.

Reviewer: The signatures are included unmodified in the block proof, not derived or modified.
```suggestion
* Generating a cryptographic block proof containing the individual node RSA signatures
  originally produced for that record file.
```

This operation ensures that all historical data remains accessible, verifiable, and compatible with
the new block stream architecture without loss of provenance or trust assumptions. While the
original record files, sidecars, and signatures will continue to be preserved in publicly
accessible cold storage, block nodes will maintain a complete historical archive of WRBs to
provide efficient, unified access to past network data within the block stream format.

> **Note:** Existing mirror nodes will not be affected by the introduction of WRBs in the block
> stream.

#### 3.2 Stage 1 — WRB Streaming Behavior

During Stage 1, consensus nodes will:

- Begin producing WRBs for each record file and streaming them to the designated Stage 1 block
  nodes.
- **Continue producing record and event streams without interruption**, regardless of the status
  of WRB production or streaming. Record stream production is not conditional on WRB success.

> - Not upload WRB based block streams to cloud storage buckets during Stage 1. Cloud bucket publishing
>   begins at Stage 2.

Reviewer: This isn't necessarily true.
The Consensus Node won't upload WRBs or block streams at any point.
Block Nodes, however, may upload WRBs to cloud storage (the 7-day retention bucket, most likely).

```suggestion
- Not upload WRBs or block streams to cloud storage buckets.
```

> - Not apply TSS signing to blocks during Stage 1. RSA-derived block proofs (as described in
>   Section 3.1) are used.

Reviewer: ```suggestion
- Not apply TSS signing to blocks during Stage 1. Block proofs containing a sufficient set of the _current_ record file signatures
  (as described in Section 3.1) are used.
```

#### 3.3 Stage 2 — Record and Event Stream Termination

At Stage 2, consensus nodes will:

- Perform a Freeze operation and stop accepting transactions.
- Freeze round comes to consensus.
- Ensure all signatures are collected and files are properly finalized.
- Produce the final record file.
- Stop producing both record and event streams at the designated cutover point.
- Shutdown uploaders as uploading responsibility is moved to block nodes.
- Process the final record file into a Wrapped Record Block (WRB). The root hash of this WRB will
  be included as the Previous Block Hash in the first block produced after the cutover, ensuring
  cryptographic continuity across the transition.

#### 3.4 Stage 2 — Block Stream Initialization

At Stage 2, consensus nodes will:

> - Construct the first block.

Reviewer: ```suggestion
- Construct the next block as a block stream.
```

- Apply the correct block number from the final record.

> - Add the root hash of the last record WRB as the Previous Block Hash.

Reviewer: ```suggestion
- Add the root hash of the last WRB as the previous block root hash.
- Continue the calculation of the all-blocks merkle subtree from the last WRB.
```

> - Sign the block using TSS and the ledger ID previously generated.

Reviewer: Not sure if we're going to make the 2 release nature of this cutover clear, but if not, we should at least be accurate that the ledger ID is _generated_ from the TSS Roster when TSS is enabled, not before then.

```suggestion
- Calculate the Ledger ID from the then-current TSS Roster
- Sign the block using TSS, the Ledger ID, and WRAPS proving key previously generated.
```

- Push block streams to the configured block nodes.

### 4\. Block Node

#### 4.1 Stage 1

During Stage 1, a limited set of designated block nodes will:

- Accept WRB streams from consensus nodes.
- Store WRBs for mirror node consumption by participating operators.
- Provide streaming access to WRBs for mirror nodes configured to receive them.
- Report ingestion metrics and errors to support readiness gate evaluation.

Block nodes not designated for Stage 1 testing are not required to participate until Stage 2.

#### 4.2 Stage 2

> At Stage 2, all block nodes will:
> 
> - Upload block streams to the new buckets following the new path structure.
> - Deduplicate block streams by block number so only one copy is stored in the bucket.
> - Provide immediate streaming access to exported consensus network data.

Reviewer: This isn't accurate, there isn't really _anything_ that all block nodes will do, block nodes won't "deduplicate" anything, and block nodes mostly won't upload blocks to cloud storage.

```suggestion
At Stage 2:

- Some block nodes will:
    - Upload block streams to the new buckets following the new path structure.
    - Only one block node will upload to each cloud so only one copy is stored in the bucket.
- Other block nodes will:
    - Store blocks locally for rapid access and long-term archive
    - Provide immediate streaming access to exported consensus network data.
```

- Testnet: `hedera-testnet-recent-block-streams`
- Previewnet: `hedera-previewnet-recent-block-streams`

> **Note:** Cloud bucket publishing does not begin during Stage 1. Buckets are populated starting

> > at Stage 2.

Reviewer: This might or might not be true; it may be better to begin populating buckets during stage 1 to ensure that process is working correctly before the whole network depends on it.

  block nodes.
- Periodically check for the existence of block stream data when no record stream data is available
  for a period of time.
- Switch from record stream processing to block stream processing permanently.
- Support retry mechanisms for failed block ingestion.
- Store block data in a compatible format with existing database schemas.

#### 6.3 Block Verification

Mirror nodes will:

- Implement block proof verification using Threshold Signature Scheme (TSS).
- Store and use the ledger ID for block verification.
- Implement retry mechanisms to handle cases where block verification fails.

> **Note:** During Stage 1, block proofs are RSA-derived (not TSS). 

#### 6.4 Bootstrap and Recovery

Mirror nodes will:

- Implement logic to process both block and record files.
- Implement logic to update the timestamp-based startup logic to use blocks rather than timestamps.
- Take PostgreSQL snapshots before the Stage 2 transition to enable mirror node bootstrapping.
- Support recovery in cases where block processing fails.

### 7\. Test Plan and Verification

#### 7.1 Stage 1 Verification

> 1. Verification that consensus nodes produce WRBs correctly alongside uninterrupted record streams.

Reviewer: ```suggestion
1. Verification that consensus nodes produce WRBs correctly alongside uninterrupted record streams.
2. Verification that WRB block proof signatures are byte-for-byte identical to corresponding record stream signatures.
```

### 8\. Disaster Recovery Considerations

#### 8.1 Stage 1 Risks

- **WRB streaming failure**: Consensus nodes must continue producing record files regardless of
  WRB streaming status. No operator action is required; record streams serve as the fallback.
- **Mirror node WRB validation failure**: Mirror nodes fall back to record streams automatically.
  Operators should monitor fallback events and report them to aid in defect diagnosis.
- **Misconfigured mirror node failing to fall back**: Mirror node configuration must be validated
  before Stage 1 to ensure fallback logic is active and correctly configured.

#### 8.2 Pre-First-Block Risks (Stage 2)

- Node failure during cutover.

> - Uploader misconfiguration (incorrect path, cloud provider, etc.).

Reviewer: ```suggestion
- Block Node cloud plugin misconfiguration (incorrect path, cloud provider, etc.).
```

- Record or event streams not stopping as intended.

#### 8.3 First Block Production Risks (Stage 2)

- Incorrect running hash.
- Incorrect block number.
- Malformed block data or schema.
- Invalid block proof (WRAPS or TSS).
- Node reconnection issues after a maintenance window.

## Backwards Compatibility

This transition introduces a fundamental change in how transaction data is exposed by a Hiero
network, but it does so in two stages to preserve compatibility throughout.

**During Stage 1**, no cryptographic or operational compatibility is broken. Record streams
continue to be produced and remain the authoritative source of network data. Mirror nodes not
configured for WRB testing are entirely unaffected. Mirror nodes configured for WRB testing retain
full fallback safety to record streams.

**At Stage 2**, the hard cryptographic transition occurs. To maintain backward compatibility:

- Historical record stream data will remain available in its current format and location for the
  foreseeable future.
- Mirror nodes will support both record stream and block stream processing.
- The first block will maintain cryptographic continuity with the last record. This includes that

>   the "previous hash" for the first block after cutover will be reported as the record running hash
>   of the last record file.

Reviewer: ```suggestion
  the "previous hash" for the first block after cutover will be reported as the WRB root hash
  of the last record file.
```

- There will be no impact on the mirror node API structure or content.

Long-term, as the ecosystem transitions to block nodes, historical record streams and cloud bucket
access may eventually be deprecated, but this would require a separate HIP and migration plan.

## Security Implications

The transition to block streams introduces several security considerations:

1. **Data Integrity**: Ensuring the running hash continuity between record streams and block

>    streams is critical to maintain the cryptographic integrity of the ledger.

Reviewer: This is tricky. The _internal_ running hashes aren't the block hashes.
The last WRB root hash will be the "previous block hash" for the first block stream block. This must be true or else we would not be able to offer state proofs for WRBs and would not have proper block chain continuity.

The record "running hash" is an internal hash (not what's signed), it's data in record stream.

Part of the reason for creating WRBs from genesis (or, roughly, OA in mainnet) is so we could produce a new continuous block chain at the level of the Block Root Hash (per HIP-1424).

```suggestion
1. **Data Integrity**: Ensuring the root hash continuity between record streams and block
   streams is critical to maintain the cryptographic integrity of the ledger.
```

2. **Block Verification**: Mirror nodes must properly implement the TSS verifier to validate block

>    proofs at Stage 2. During Stage 1, RSA-derived proof verification is used and must also be

Reviewer: RSA-derived suggests we're doing something to change the signatures.
We're just using the existing signatures we currently validate.  The only change is that they're gathered together in a `BlockProof` protobuf instead of individual `.sig` files.

Author: So you suggesting to remove the 'derived' from the text?

Reviewer: Yes, I would remove the word derived.

Alternatively, replace the clause as follows:

```suggestion
   proofs at Stage 2. During Stage 1, existing RSA signature verification is used and must also be
```

   correctly implemented.

3. **Stage 1 Authority Boundary**: During Stage 1, WRBs are not cryptographically authoritative.
   Mirror nodes and consumers must not treat Stage 1 WRBs as a substitute for record stream

>    finality. The record stream remains the sole source of truth during this stage.

Reviewer: This isn't accurate, and it's confusing.
WRBs _are_ the record stream and have the same signatures. As such, they _are_ authoritative and can be used interchangeably with the record files. If this is not true, then the WRB history must be recreated. WRBs only value is that continuity and interchangeability.

```suggestion
3. **Stage 1 Authority Boundary**: During Stage 1, WRBs contain identical content and share the same
   cryptographic authority as record files. Mirror nodes and consumers _may_ treat Stage 1 WRBs
   as a substitute for record stream finality, but are not required to do so.
```

Note, Because CN's must gather signatures under time pressure, each WRB contains _enough_ signatures to be validated and ensure no network partition, but may not contain _all_ signatures. From a network security perspective, this is not relevant, but if a client wants _all_ signatures, they should prefer the cloud storage while it is available.

2. **For Developers:**

    - Explain the conceptual shift from records to blocks.
    - Clarify that Stage 1 does not change API behavior or data availability for downstream
      consumers.

>     - Document how transaction data will be organized in the new format.

Reviewer: ```suggestion
    - Emphasize that the Mirror Node behavior and APIs remains unchanged
    - Emphasize that the Consensus Node HAPI interface is unchanged.
    - Document how transaction data will be organized in the new format.
```

    - Provide examples of how to process block data.

3. **For Node Operators:**

    - Detail the changes required to consensus node configuration for each stage.

>     - Explain the new uploader paths and configurations.

Reviewer: There are no new uploader paths or configurations.
CN doesn't upload anything new.

Author: There will be block streaming configuration on the CN at both stages, no?

Reviewer: The commented line is
`    - Explain the new uploader paths and configurations.`

The only explanation here would be removal of the old record file paths in the stage 2 upgrade, nothing about new paths _for consensus node operators_.
